### PR TITLE
Fully Enable CLOG (LTTng) for Linux

### PR DESCRIPTION
**Pull request by chgray**
This change updates the Linux builds to fully enable LTTng based CLOG support. Windows is currently left alone.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -393,4 +392,4 @@ if(QUIC_BUILD_TEST)
     add_subdirectory(src/platform/unittest)
     add_subdirectory(src/test/lib)
     add_subdirectory(src/test/bin)
-endif()
+endif()
```
Comment (nibanks) [inline CMakeLists.txt:395]: nit: add newline back

File: scripts/build.ps1 [hunk 2]
```diff
@@ -152,6 +152,8 @@ if ("" -eq $Platform) {
     }
 }
 
+$env:CLOG_DEVELOPMENT_MODE = 1
```
Comment (nibanks) [inline scripts/build.ps1:155]: This is not supposed to be checked in.

File: src/test/lib/ApiTest.cpp [hunk 3]
```diff
@@ -23,6 +23,8 @@ void QuicTestValidateApi()
         MsQuicOpen(nullptr));
 
     MsQuicClose(nullptr);
+
+    TEST_FAILURE("Forcing test failure to test CLOG"); // TODO - Remove before merging feature/clog to master
```
Comment (nibanks) [inline src/test/lib/ApiTest.cpp:27]: Remove this now.

File: src/core/binding.c [hunk 4]
```diff
@@ -146,7 +146,7 @@ QuicBindingInitialize(
     QUIC_ADDR DatapathLocalAddr, DatapathRemoteAddr;
     QuicDataPathBindingGetLocalAddress(Binding->DatapathBinding, &DatapathLocalAddr);
     QuicDataPathBindingGetRemoteAddress(Binding->DatapathBinding, &DatapathRemoteAddr);
-    QuicTraceEvent(
+    QuicTraceEvent_Skip(
```
Comment (nibanks) [inline src/core/binding.c:149]: I want to not need this. Let's finish this up so that it's not necessary.

File: scripts/build.ps1 (hunk 2 above)
Comment (ThadHouse) [reply scripts/build.ps1:155]: I know, but it fails in CI and I just want to see if it builds at all.

Comment (nibanks) [reply scripts/build.ps1:155]: If it does build, it means we need to rebuild locally, to update the sidecar, and then the CI won't need it.

Comment (ThadHouse) [reply scripts/build.ps1:155]: One flaw in this I noticed is runs on windows don't seem to affect linux. So if we ever updated, we'd have to run manually on all platforms. Not too difficult until we add mac.

Comment (nibanks) [reply scripts/build.ps1:155]: Perhaps a special pipeline that runs all platforms up update the config for us? Or have that automatically happen for PR CI runs?

File: src/core/binding.c [hunk 5]
```diff
@@ -151,8 +151,8 @@ QuicBindingInitialize(
         "[bind][%p] Created, Udp=%p LocalAddr=%!SOCKADDR! RemoteAddr=%!SOCKADDR!",
         Binding,
         Binding->DatapathBinding,
-        LOG_BINARY(sizeof(DatapathLocalAddr), &DatapathLocalAddr),
-        LOG_BINARY(sizeof(DatapathRemoteAddr), &DatapathRemoteAddr));
+        CLOG_BYTEARRAY(sizeof(DatapathLocalAddr), &DatapathLocalAddr),
```
Comment (nibanks) [inline src/core/binding.c:154]: One benefit of doing something more generic, like `LOG_BINARY` is if we have to pull the plug in CLOG, we don't have to do a global find/replace.

File: src/core/CMakeLists.txt [hunk 6]
```diff
@@ -44,7 +44,7 @@ if(NOT "${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
 endif()
 
 # Allow CLOG to preprocess all the source files.
-add_clog_library(core.clog STATIC ${SOURCES})
+add_clog_library(core.clog SHARED ${SOURCES})
```
Comment (nibanks) [inline src/core/CMakeLists.txt:47]: Why shared? We got things working with a static LTTNg just fine. Also, we definitely don't want shared for Windows (when we get there).

File: src/core/connection.c [hunk 7]
```diff
@@ -3753,13 +3753,15 @@ QuicConnRecvDecryptAndAuthenticate(
             Packet->HeaderLength);
     }
 
-    QuicTraceEvent(
+#if 0
```
Comment (nibanks) [inline src/core/connection.c:3756]: What problem are you having with this event?

File: src/manifest/msquic.clog_config [hunk 8]
```diff
@@ -75,6 +79,138 @@
...
+      }
+    },
+    {
+      "MacroName": "QuicTraceLogError",
+      "EncodedPrefix": null,
+      "EncodedArgNumber": 1,
+      "MacroConfiguration": {
+        "linux": "lttng_plus",
+       "stubs": "stubs",
+        "windows_kernel": "empty",
```
Comment (nibanks) [inline src/manifest/msquic.clog_config:155]:
```suggestion
        "linux": "lttng_plus",
        "stubs": "stubs",
        "windows_kernel": "empty",
```

File: src/manifest/msquic.clog_config [hunk 9]
```diff
@@ -75,6 +79,138 @@
     ]
   },
   "SourceCodeMacros": [
+    {
+      "MacroName": "QuicTraceLogWarning",
+      "EncodedPrefix": null,
+      "EncodedArgNumber": 1,
+      "MacroConfiguration": {
+        "linux": "lttng_plus",
```
Comment (nibanks) [inline src/manifest/msquic.clog_config:87]: I feel like we should rename `lttng_plus` back to just `lttng` since we're making this official.

**Comment by nibanks** [comment]
Looks like the build is failing for Windows and Linux arm (Raspberry Pi). Also, there are a lot of extra output lines in the Windows build:
```
  Skipping Macro : QuicTraceLogWarning due to 'SkipProcessing' being set
  Skipping Macro : QuicTraceLogStreamWarning due to 'SkipProcessing' being set
  Skipping Macro : QuicTraceLogStreamInfo due to 'SkipProcessing' being set
  Skipping Macro : QuicTraceLogStreamVerbose due to 'SkipProcessing' being set
  Skipping Macro : QuicTraceLogInfo due to 'SkipProcessing' being set
  Skipping Macro : QuicTraceLogVerbose due to 'SkipProcessing' being set
  Skipping Macro : QuicTraceLogError due to 'SkipProcessing' being set
  Skipping Macro : QuicTraceLogConnError due to 'SkipProcessing' being set
  Skipping Macro : QuicTraceLogConnWarning due to 'SkipProcessing' being set
  Skipping Macro : QuicTraceLogConnInfo due to 'SkipProcessing' being set
  Skipping Macro : QuicTraceLogConnVerbose due to 'SkipProcessing' being set
  Skipping Macro : QuicTraceEvent due to 'SkipProcessing' being set
```
I think we should get rid of these.

**Comment by ThadHouse** [comment]
Once the dynamic tracepoint provider support is added, every platform should be able to use static.

File: src/inc/quic_trace.h [hunk 10]
```diff
@@ -243,6 +250,7 @@ QuicTraceStubVarArgs(
         EventWriteQuic##Type##Log##EventName##_AssumeEnabled(Ptr, EtwBuffer); \
     }
 
+
```
Comment (nibanks) [inline src/inc/quic_trace.h:253]: OMG Chris! Do you just close your eyes, click somewhere and hit Enter?!

File: src/core/connection.c [hunk 11]
```diff
@@ -3753,13 +3753,15 @@ QuicConnRecvDecryptAndAuthenticate(
...
+    uicTraceEvent(
```
Comment (nibanks) [inline src/core/connection.c:3757]: Let's fix this before merging.

**Pull request merged**